specs/core/1_custody-game.md

```diff
@@ -263,45 +275,99 @@ class BeaconBlockBody(Container):
 
 ```python
 def ceillog2(x: uint64) -> int:
-    return x.bit_length()
+    return (x - 1).bit_length()
+```
+
+### `is_valid_merkle_branch_with_mixin`
+
+```python
+def is_valid_merkle_branch_with_mixin(leaf: Hash, 
+                                      branch: Sequence[Hash],
+                                      depth: uint64,
+                                      index: uint64,
+                                      root: Hash,
+                                      mixin: uint64) -> bool:
+    value = leaf
+    for i in range(depth):
+        if index // (2**i) % 2:
+            value = hash(branch[i] + value)
+        else:
+            value = hash(value + branch[i])
+    value = hash(value + mixin.to_bytes(32, "little"))
+    return value == root
 ```
 
 ### `get_crosslink_chunk_count`
 
 ```python
 def get_custody_chunk_count(crosslink: Crosslink) -> int:
     crosslink_length = min(MAX_EPOCHS_PER_CROSSLINK, crosslink.end_epoch - crosslink.start_epoch)
-    chunks_per_epoch = 2 * BYTES_PER_SHARD_BLOCK * SLOTS_PER_EPOCH // BYTES_PER_CUSTODY_CHUNK
-    return crosslink_length * chunks_per_epoch
+    return crosslink_length * CHUNKS_PER_EPOCH
 ```
 
-### `get_bit`
+### `legendre_bit`
+
+Returns the Legendre symbol `(a/q)` normalizes as a bit (i.e. `((a/q) + 1) // 2`). In a production implementation, a well-optimized library (e.g. GMP) should be used for this.
 
 ```python
-def get_bit(serialization: bytes, i: uint64) -> int:
-    """
-    Extract the bit in ``serialization`` at position ``i``.
-    """
-    return (serialization[i // 8] >> (i % 8)) % 2
+def legendre_bit(a: int, q: int) -> int:
+    if a >= q:
+        return legendre_bit(a % q, q)
+    if a == 0:
+        return 0 
+    assert(q > a > 0 and q % 2 == 1)
+    t = 1
+    n = q
+    while a != 0:
+        while a % 2 == 0:
+            a //= 2
+            r = n % 8
+            if r == 3 or r == 5:
+                t = -t
+        a, n = n, a
+        if a % 4 == n % 4 == 3:
+            t = -t
+        a %= n
+    if n == 1:
+        return (t + 1) // 2
+    else:
+        return 0
+```
+
+### ```custody_subchunkify```
+
+Given one proof of custody chunk, returns the proof of custody subchunks of the correct sizes.
+
+```python
+def custody_subchunkify(x: bytes) -> list:
+    subchunks = []
+    start = 0
+    for size in BYTES_PER_CUSTODY_SUBCHUNK:
+        subchunks.append(x[start:start + size])
+        start += size
+    return subchunks
 ```
 
 ### `get_custody_chunk_bit`
 
 ```python
 def get_custody_chunk_bit(key: BLSSignature, chunk: bytes) -> bool:
-    # TODO: Replace with something MPC-friendly, e.g. the Legendre symbol
-    return bool(get_bit(hash(key + chunk), 0))
+    full_G2_element = bls_signature_to_G2(key)
+    s = full_G2_element[0].coeffs
+    bits = [legendre_bit((i + 1) * s[i % 2] + int.from_bytes(subchunk, "little"), BLS12_381_Q)
+            for i, subchunk in enumerate(custody_subchunkify(chunk))]
+
+    return bool(sum(bits) % 2)
```

Comment:
Cryptography looks good here. For a moment I thought it was broken because you could "Factor out" the (i+1) and (s[i+2]) and/or combine all the subchunks together, but I saw that you're adding the chunk and the key rather than multiplying, and I don't think there's additive properties of Legendre that could be attacked here.

Reply:
So you can actually multiply all the subchunks together, which I consider a feature -- doing that improves computation time by a factor of 3 in an MPC (and I think actually a similar amount in a cleartext computation). I think it's not a problem because for oursourcing, the outsourcer would still have to give away a multivariate polynomial of which the roots are his secrets `s[0]` and `s[1]`.